### PR TITLE
docs - updates for online expand

### PR DESCRIPTION
**Pull request by mkiyama**
These are changes for the HA "how to" (procedural) topics in the Admin Guide for current online expand.
--Removed mention of system being offline during new segment initialization
--Add information about table redistribution using rebuild and move
--Other misc. updates.

Changes to reference information
--Add GUCs gp_expand_method and gp_use_legacy_hashops
--Changes to catalog tables.
--updates to gpexpand utility

TO DO
--need to test w/ working utilities when available.
--update docs when gpexpand online feature branch is merged.

Link to HTML format of updated docs on a GPDB review site.
Admin Guide topics
https://docs-msk-gpdb6-dev.cfapps.io/600/admin_guide/expand/expand-initialize.html
https://docs-msk-gpdb6-dev.cfapps.io/600/admin_guide/expand/expand-main.html
https://docs-msk-gpdb6-dev.cfapps.io/600/admin_guide/expand/expand-nodes.html
https://docs-msk-gpdb6-dev.cfapps.io/600/admin_guide/expand/expand-overview.html
https://docs-msk-gpdb6-dev.cfapps.io/600/admin_guide/expand/expand-planning.html
https://docs-msk-gpdb6-dev.cfapps.io/600/admin_guide/expand/expand-redistribute.html

Util Guide topic
https://docs-msk-gpdb6-dev.cfapps.io/600/utility_guide/admin_utilities/gpexpand.html

New GUCs gp_expand_method, gp_use_legacy_hashops
https://docs-msk-gpdb6-dev.cfapps.io/600/ref_guide/config_params/guc-list.html#gp_expand_method
https://docs-msk-gpdb6-dev.cfapps.io/600/ref_guide/config_params/guc-list.html#gp_use_legacy_hashops

Changes to catalog tables
https://docs-msk-gpdb6-dev.cfapps.io/600/ref_guide/system_catalogs/gp_distribution_policy.html
https://docs-msk-gpdb6-dev.cfapps.io/600/ref_guide/system_catalogs/gp_expansion_tables.html
https://docs-msk-gpdb6-dev.cfapps.io/600/ref_guide/system_catalogs/pg_tablespace.html

### CODE REVIEW

**Comment by gaos1** [comment]
Thanks @mkiyama. I add @boogabee to reviewers too.

File: gpdb-doc/dita/admin_guide/expand/expand-planning.xml [hunk 1]
```diff
@@ -217,33 +225,51 @@
...
         factors related to hardware performance. In most environments, the initialization of new
         segments requires less than thirty minutes offline.</p>
-      <note type="note">After you begin initializing new segments, you can no longer restore the
-        system using backup files created for the pre-expansion system. When initialization
+      <p>These utilities cannot be run while <codeph>gpexpand</codeph> is performing segment
+          initialization.<ul id="ul_c3w_wdp_lgb">
+          <li><codeph>gppkg</codeph></li>
+          <li><codeph>gpconfig</codeph></li>
+          <li><codeph>gpcheck</codeph></li>
+          <li><codeph>gpcheckat</codeph></li>
```
Comment (pivotal-ning-yu) [inline gpdb-doc/dita/admin_guide/expand/expand-planning.xml]: We are still discussing whether `gpcheckcat` should be conflict with gpexpand, will update here once we got the conclusion.

File: gpdb-doc/dita/admin_guide/expand/expand-initialize.xml [hunk 2]
```diff
@@ -151,18 +153,18 @@ sdw5:sdw5-1:60012:/gpdata/mirror/gp10:14:10:m</codeblock>
               <row>
                 <entry colname="col1">datadir</entry>
                 <entry colname="col2">Directory name</entry>
-                <entry colname="col3">The data directory location for a segment as per
-                  the gp_segment_configuration system catalog.</entry>
+                <entry colname="col3">The data directory location for a segment as per the
+                  gp_segment_configuration system catalog.</entry>
```
Comment (dyozie) [inline gpdb-doc/dita/admin_guide/expand/expand-initialize.xml]: add codeph tag to catalog table name here.

File: gpdb-doc/dita/admin_guide/expand/expand-redistribute.xml [hunk 3]
```diff
@@ -77,17 +85,19 @@
...
+            >gpexpand.status_detail</xref></i> for per-table status information.</p>
     </body>
     <topic id="topic32" xml:lang="en">
       <title>Viewing Expansion Status</title>
       <body>
         <p>After the first table completes redistribution, <i>gpexpand.expansion_progress</i>
           calculates its estimates and refreshes them based on all tables' redistribution rates.
           Calculations restart each time you start a table redistribution session with
-            <codeph>gpexpand</codeph>. To monitor progress, connect to Greenplum Database using
+            <codeph>gpexpand</codeph>. To monitor progress, connect to Greenplum Data\base using
```
Comment (dyozie) [inline gpdb-doc/dita/admin_guide/expand/expand-redistribute.xml]: Stray character here:  Data\base

File: gpdb-doc/dita/admin_guide/expand/expand-redistribute.xml [hunk 4]
```diff
@@ -34,13 +38,15 @@
...
-        tables and minimize performance impact. Available free disk space can affect table ranking;
-        see <xref href="expand-planning.xml#topic11" type="topic" format="dita"/>.</p>
-      <p>To rank tables for redistribution by updating <codeph>rank</codeph> values in
-          <i>gpexpand.status_detail</i>, connect to Greenplum Database using <codeph>psql</codeph>
-        or another supported client. Update <i>gpexpand.status_detail</i> with commands such as:</p>
+      <p>For large systems, you can control the table redistribution order. Adjust tables'
+          <codeph>rank</codeph> values in the expansion schema to prioritize heavily-used tables and
+        minimize performance impact. Available free disk space can affect table ranking; see <xref
+          href="expand-planning.xml#topic11" type="topic" format="dita"/>.</p>
+      <p>To rank tables for redistribution by updating <codeph>rank</codeph> values in <i><xref
```
Comment (dyozie) [inline gpdb-doc/dita/admin_guide/expand/expand-redistribute.xml]: The table name here and in a few other spots is italicized instead of codeph format.

File: gpdb-doc/dita/ref_guide/system_catalogs/gp_expansion_tables.xml [hunk 5]
```diff
@@ -4,6 +4,8 @@
 <topic id="topic1" xml:lang="en">
   <title id="fb138336">gpexpand.status_detail</title>
   <body>
+    <draft-comment author="msk">updates for GPDB 6<p>distribution_policy_type,
+        root_partition_name</p></draft-comment>
```
Comment (dyozie) [inline gpdb-doc/dita/ref_guide/system_catalogs/gp_expansion_tables.xml]: Not part of this PR, but there's a typo (two periods) at the end of the final intro paragraph in https://docs-msk-gpdb6-dev.cfapps.io/600/ref_guide/system_catalogs/gp_expansion_tables.html

File: gpdb-doc/dita/ref_guide/config_params/guc-list.xml [hunk 6]
```diff
@@ -3347,6 +3351,51 @@
...
+  <topic id="gp_expand_method">
+    <title>gp_expand_method</title>
+    <body>
+      <draft-comment author="msk">DRAFT <xref
+          href="https://www.pivotaltracker.com/story/show/162535111" format="html" scope="external"
+          >162535111</xref>
+        <xref href="https://github.com/greenplum-db/gpdb/pull/6630" format="html" scope="external"
+          >gpdb-feature/pull/66300</xref> merged to feature<p><b>[Q:]</b> Will both expand methods
+          work with both hashmethods?</p></draft-comment>
+      <p>This parameter controls how<codeph>gpexpand</codeph> redistributes table data during the
```
Comment (dyozie) [inline gpdb-doc/dita/ref_guide/config_params/guc-list.xml]: missing space after "how"

File: gpdb-doc/dita/ref_guide/config_params/guc-list.xml [hunk 7]
```diff
@@ -3347,6 +3351,51 @@
...
+        table data redistribution phase of a system expansion. There are two ways
+          <codeph>gpexpand</codeph> can perform table data redistribution.<ul id="ol_xdd_1py_jgb">
+          <li>rebuild - Create a new table, copy all the data from the old to the new table, and
```
Comment (dyozie) [inline gpdb-doc/dita/ref_guide/config_params/guc-list.xml]: Probably best to format "rebuild" and "move" as codeph to indicate that they're the literal option names.

File: gpdb-doc/dita/ref_guide/config_params/guc-list.xml [hunk 8]
```diff
@@ -3347,6 +3351,51 @@
...
+            replace the old table. This is the default. The rebuild method is similar creating a new
```
Comment (dyozie) [inline gpdb-doc/dita/ref_guide/config_params/guc-list.xml]: similar creating -> similar to creating

File: gpdb-doc/dita/ref_guide/system_catalogs/gp_distribution_policy.xml [hunk 9]
```diff
@@ -25,31 +27,37 @@
...
-                policy.</li></ul></entry>
+                <li><codeph>p</codeph> - The table has a partitioned distribution policy.</li>
+                <li><codeph>r</codeph> - The table has a replicated distribution policy.</li>
+              </ul></entry>
+          </row>
+          <row>
+            <entry colname="col1"><codeph>numsegments</codeph></entry>
+            <entry colname="col2">integer</entry>
+            <entry colname="col3"/>
+            <entry colname="col4">The number of segment instances the table data is distributed
```
Comment (dyozie) [inline gpdb-doc/dita/ref_guide/system_catalogs/gp_distribution_policy.xml]: the table data is distributed on -> on which the table data is distributed

File: gpdb-doc/dita/admin_guide/expand/expand-overview.xml [hunk 10]
```diff
@@ -48,43 +50,43 @@
...
-            <li>Databases and database objects are created on the new segment hosts. </li>
-            <li>An expansion schema is created in the master database to control the expansion
-              process.</li>
-            <li>The distribution policy for each table is changed to <codeph>DISTRIBUTED
-                RANDOMLY</codeph>.</li>
-          </ul></li>
-        <li>The system is restarted and applications resume. <ul id="ul_tmw_b3h_mr">
+            <li>Databases and database objects are created in the new segment instances on the new
+              segment hosts. </li>
+            <li>The <i>gpexpand</i> schema is created in the master database. You use the tables in
```
[the remark this answers is not in the document]
Comment (mkiyama) [reply gpdb-doc/dita/admin_guide/expand/expand-overview.xml]: you can use gpexpand.status_detail to control the order that the tables are redistributed to the segments.

**Comment by pivotal-ning-yu** [comment]
On Thu, Jan 24, 2019 at 7:29 AM Mel Kiyama <notifications@github.com> wrote:

> *@mkiyama* commented on this pull request.
> ------------------------------
>
> In gpdb-doc/dita/admin_guide/expand/expand-planning.xml
> <https://github.com/greenplum-db/gpdb/pull/6719#discussion_r250414186>:
>
> > +              <li>rebuild - Create a new table, copy all the data from the old to the new table, and
> +                replace the old table. This is the default. The rebuild method is similar creating a
> +                new table with a <codeph>CREATE TABLE AS SELECT</codeph> command.</li>
> +              <li>move - Scan all the data and perform an <codeph>UPDATE</codeph> operation to move
> +                rows as needed to different segment instances. In general, this method requires less
> +                disk space, however, it creates obsolete table rows and might require a
> +                  <codeph>VACUUM</codeph> operation on the table. Also, this method updates indexes
> +                one row at a time, which can be much slower than rebuilding the index with the
> +                  <codeph>CREATE INDEX</codeph> command.</li>
> +            </ul></p>
> +          <p>The server configuration parameter <codeph><xref
> +                href="../../ref_guide/config_params/guc-list.xml#gp_expand_method"
> +                >gp_expand_method</xref></codeph> controls the redistribution method.</p>
> +          <draft-comment author="msk">leave out until gp_use_legacy_hashops and upgrade are
> +            available.</draft-comment>
> +          <p>When planning the redistribution phase, consider the impact of the exclusive lock taken
>
> @gaos1 <https://github.com/gaos1> @pivotal-ning-yu
> <https://github.com/pivotal-ning-yu>
>
> This paragraph is from GPDB 5 docs. Can you clarify, what type of locking
> is used during the redistribution phase for rebuild and move?
>

In gpdb5, as well as current gpdb6, ALTER TABLE t1 SET WITH
(reorganize=true) will acquire AccessExclusiveLock, the highest one, on t1,
this will block all concurrent select, dml and ddl on the table.

"rebuild" is the same as above command, "move" is only in gpdb6, both
acquire AccessExclusiveLock at the moment.

When planning the redistribution phase, consider the impact of the
> exclusive lock taken on each table during redistribution. User activity on
> a table can delay its redistribution, but also tables are unavailable for
> user activity during redistribution.
>

Yes this is true, user activity and redistribution on the same table are
serialized, so the redistribution must be carefully planned and scheduled.

**Comment by dyozie** [comment]
Merging this in now with the feedback to-date.  But feel free to continue reviewing/commenting. Further work will occur in new PRs.
